app/background.cc

```diff
@@ -0,0 +1,161 @@
+/* This is licensed software, @see LICENSE file.
+ * Authors - @see AUTHORS file. */
+
+#include <stdio.h>
+#include <time.h>
+#include <unistd.h>
+#include <thread>
+#include <mutex>
+#include <opencv2/videoio.hpp>
+#include <opencv2/imgproc.hpp>
+#include <opencv2/highgui.hpp>
+
+// Internal state of background processing
+struct background_t {
+    int dbg;
+    bool vid;
+    volatile bool run;
+    cv::VideoCapture cap;
+    int frm;
+    double fps;
+    cv::Mat raw;
+    std::thread thr;
+    std::mutex mux;
+};
+
+// Internal video reader thread
+static void read_thread(background_t *pbkd) {
+    if (pbkd->dbg) fprintf(stderr, "background: thread start\n");
+    struct timespec last;
+    struct timespec proc;
+    clock_gettime(CLOCK_REALTIME, &last);
+    proc = last;
+    while (pbkd->run) {
+        // read frame under lock
+        pbkd->mux.lock();
```

Comment:
Never unlocked if read feils

Reply:
Also use RAII locks via std::lock_guard here if possible.

Reply:
Yep, spotted that - fell in my own trap! Now using RAII locks.